facebook/OpenBIC
# fby3.5: cl: Support I3C read PMIC critical error

Summary:
- Read PMIC critical error through I3C when host shutdown.

Test plan:
- Build code: Pass
- Read PMIC critical through I3C:
DIMM                              PMIC              Result
Micron MTC10F1084S1RC48BA1 PHCC   Richtek RTQ5119   Failed
SK Hynix HMCG78MEBRA113N AA       TI_TPS53832 178   Failed
Micron MTC10F1084S1RC48BA1 JCCI   TI_TPS53832 188   Failed
SK Hynix HMCG78MEBRA174N AA       Renesas P8910-Z1  Pass

Log:
1. Check all PMIC can read critical through I3C. root@bmc-oob:~# power-util slot1 status
Power status for fru 1 : ON
root@bmc-oob:~# dimm-util slot1 --pmic --dimm A0 --err_inj VinB_OV root@bmc-oob:~# power-util slot1 status
Power status for fru 1 : OFF
root@bmc-oob:~# log-util --print slot1
1    slot1    2022-10-24 04:22:13    ipmid            SEL Entry: FRU: 1, Record: Standard (0x02), Time: 2022-10-24 04:22:13, Sensor:      PMIC_ERR (0xB4), Event Data: (000400) DIMM A0 Vin Bulk OV Assertion
1    slot1    2022-10-24 04:22:16    gpiod            FRU: 1, System powered OFF

root@bmc-oob:~# power-util slot1 status
Power status for fru 1 : ON
root@bmc-oob:~# dimm-util slot1 --pmic --dimm A2 --err_inj VinB_OV root@bmc-oob:~# power-util slot1 status
Power status for fru 1 : OFF
root@bmc-oob:~# log-util --print slot1
1    slot1    2022-10-24 04:24:59    ipmid            SEL Entry: FRU: 1, Record: Standard (0x02), Time: 2022-10-24 04:24:59, Sensor: PMIC_ERR (0xB4), Event Data: (010400) DIMM A2 Vin Bulk OV Assertion
1    slot1    2022-10-24 04:25:02    gpiod            FRU: 1, System powered OFF

root@bmc-oob:~# power-util slot1 status
Power status for fru 1 : ON
root@bmc-oob:~# dimm-util slot1 --pmic --dimm A3 --err_inj VinB_OV root@bmc-oob:~# power-util slot1 status
Power status for fru 1 : OFF
root@bmc-oob:~# log-util --print slot1
1    slot1    2022-10-24 04:28:40    ipmid            SEL Entry: FRU: 1, Record: Standard (0x02), Time: 2022-10-24 04:28:40, Sensor: PMIC_ERR (0xB4), Event Data: (020400) DIMM A3 Vin Bulk OV Assertion
1    slot1    2022-10-24 04:28:43    gpiod            FRU: 1, System powered OFF

root@bmc-oob:~# power-util slot1 status
Power status for fru 1 : ON
root@bmc-oob:~# dimm-util slot1 --pmic --dimm A4 --err_inj VinB_OV root@bmc-oob:~# power-util slot1 status
Power status for fru 1 : OFF
root@bmc-oob:~# log-util --print slot1
1    slot1    2022-10-24 04:31:34    ipmid            SEL Entry: FRU: 1, Record: Standard (0x02), Time: 2022-10-24 04:31:34, Sensor: PMIC_ERR (0xB4), Event Data: (030400) DIMM A4 Vin Bulk OV Assertion
1    slot1    2022-10-24 04:31:37    gpiod            FRU: 1, System powered OFF

root@bmc-oob:~# power-util slot1 status
Power status for fru 1 : ON
root@bmc-oob:~# dimm-util slot1 --pmic --dimm A6 --err_inj VinB_OV root@bmc-oob:~# power-util slot1 status
Power status for fru 1 : OFF
root@bmc-oob:~# log-util --print slot1
1    slot1    2022-10-24 04:34:49    ipmid            SEL Entry: FRU: 1, Record: Standard (0x02), Time: 2022-10-24 04:34:49, Sensor: PMIC_ERR (0xB4), Event Data: (040400) DIMM A6 Vin Bulk OV Assertion
1    slot1    2022-10-24 04:34:52    gpiod            FRU: 1, System powered OFF

root@bmc-oob:~# power-util slot1 status
Power status for fru 1 : ON
root@bmc-oob:~# dimm-util slot1 --pmic --dimm A7 --err_inj VinB_OV root@bmc-oob:~# power-util slot1 status
Power status for fru 1 : OFF
root@bmc-oob:~# log-util --print slot1
1    slot1    2022-10-24 04:38:53    ipmid            SEL Entry: FRU: 1, Record: Standard (0x02), Time: 2022-10-24 04:38:53, Sensor: PMIC_ERR (0xB4), Event Data: (050400) DIMM A7 Vin Bulk OV Assertion
1    slot1    2022-10-24 04:38:57    gpiod            FRU: 1, System powered OFF

2. Check different critical errors can be read.
- SWAout_OV root@bmc-oob:~# power-util slot1 status
Power status for fru 1 : ON
root@bmc-oob:~# dimm-util slot1 --pmic --dimm A0 --err_inj SWAout_OV root@bmc-oob:~# power-util slot1 status
Power status for fru 1 : OFF
root@bmc-oob:~# log-util --print slot1
1    slot1    2022-10-24 05:01:28    ipmid            SEL Entry: FRU: 1, Record: Standard (0x02), Time: 2022-10-24 05:01:28, Sensor: PMIC_ERR (0xB4), Event Data: (000000) DIMM A0 SWAout OV Assertion
1    slot1    2022-10-24 05:01:31    gpiod            FRU: 1, System powered OFF

- SWCout_OV root@bmc-oob:~# power-util slot1 status
Power status for fru 1 : ON
root@bmc-oob:~# dimm-util slot1 --pmic --dimm A0 --err_inj SWCout_OV root@bmc-oob:~# power-util slot1 status
Power status for fru 1 : OFF
root@bmc-oob:~# log-util --print slot1
1    slot1    2022-10-24 05:03:49    ipmid            SEL Entry: FRU: 1, Record: Standard (0x02), Time: 2022-10-24 05:03:49, Sensor: PMIC_ERR (0xB4), Event Data: (000200) DIMM A0 SWCout OV Assertion
1    slot1    2022-10-24 05:03:51    gpiod            FRU: 1, System powered OFF

- SWDout_OV root@bmc-oob:~# power-util slot1 status
Power status for fru 1 : ON
root@bmc-oob:~# dimm-util slot1 --pmic --dimm A0 --err_inj SWDout_OV root@bmc-oob:~# power-util slot1 status
Power status for fru 1 : OFF
root@bmc-oob:~# log-util --print slot1
1    slot1    2022-10-24 05:07:30    ipmid            SEL Entry: FRU: 1, Record: Standard (0x02), Time: 2022-10-24 05:07:30, Sensor: PMIC_ERR (0xB4), Event Data: (000300) DIMM A0 SWDout OV Assertion
1    slot1    2022-10-24 05:07:33    gpiod            FRU: 1, System powered OFF

- SWAout_UV root@bmc-oob:~# power-util slot1 status
Power status for fru 1 : ON
root@bmc-oob:~# dimm-util slot1 --pmic --dimm A0 --err_inj SWAout_UV root@bmc-oob:~# power-util slot1 status
Power status for fru 1 : OFF
root@bmc-oob:~# log-util --print slot1
1    slot1    2022-10-24 05:10:10    ipmid            SEL Entry: FRU: 1, Record: Standard (0x02), Time: 2022-10-24 05:10:10, Sensor: PMIC_ERR (0xB4), Event Data: (000600) DIMM A0 SWAout UV Assertion
1    slot1    2022-10-24 05:10:13    gpiod            FRU: 1, System powered OFF

- SWCout_UV root@bmc-oob:~# power-util slot1 status
Power status for fru 1 : ON
root@bmc-oob:~# dimm-util slot1 --pmic --dimm A0 --err_inj SWCout_UV root@bmc-oob:~# power-util slot1 status
Power status for fru 1 : OFF
root@bmc-oob:~# log-util --print slot1
1    slot1    2022-10-24 05:12:22    ipmid            SEL Entry: FRU: 1, Record: Standard (0x02), Time: 2022-10-24 05:12:22, Sensor: PMIC_ERR (0xB4), Event Data: (000800) DIMM A0 SWCout UV Assertion
1    slot1    2022-10-24 05:12:25    gpiod            FRU: 1, System powered OFF

- SWDout_UV root@bmc-oob:~# power-util slot1 status
Power status for fru 1 : ON
root@bmc-oob:~# dimm-util slot1 --pmic --dimm A0 --err_inj SWDout_UV root@bmc-oob:~# power-util slot1 status
Power status for fru 1 : OFF
root@bmc-oob:~# log-util --print slot1
1    slot1    2022-10-24 05:14:03    ipmid            SEL Entry: FRU: 1, Record: Standard (0x02), Time: 2022-10-24 05:14:03, Sensor: PMIC_ERR (0xB4), Event Data: (000900) DIMM A0 SWDout UV Assertion
1    slot1    2022-10-24 05:14:06    gpiod            FRU: 1, System powered OFF

- Critical_temp_shutdown root@bmc-oob:~# power-util slot1 status
Power status for fru 1 : ON
root@bmc-oob:~# dimm-util slot1 --pmic --dimm A0 --err_inj critical_temp_shutdown root@bmc-oob:~# power-util slot1 status
Power status for fru 1 : OFF
root@bmc-oob:~# log-util --print slot1
1    slot1    2022-10-24 05:16:00    ipmid            SEL Entry: FRU: 1, Record: Standard (0x02), Time: 2022-10-24 05:16:00, Sensor: PMIC_ERR (0xB4), Event Data: (001000) DIMM A0 Current temp shutdown Assertion
1    slot1    2022-10-24 05:16:02    gpiod            FRU: 1, System powered OFF
